### PR TITLE
[PROF-2710] Fix missing instrumentation when using Thread.start/fork

### PR DESCRIPTION
**Pull request by ivoanjo**
When a `Thread` is started with either `start` or `fork` (they are aliases), then our `CThread#initialize` patch never gets run and thus these threads are not properly instrumented.

This API is used in Ruby's own built-in `timeout` gem (see <https://github.com/ruby/timeout/blob/14d421252b54027108da183783a7a26d0cf5aac1/lib/timeout.rb#L87>) although I'm not entirely sure how common it is in the community.

The behavior of skipping `#initialize` is by design, but it seems a rather weird API; here's what the docs say:

> Basically the same as ::new. However, if class Thread is subclassed,
> then calling start in that subclass will not invoke the subclass's
> initialize method.

(<https://rubyapi.org/3.0/o/thread#method-c-start>)

To fix this, we introduce a new module, which wraps these two operations to provide the expected behavior of installing the CPU instrumentation.

**NOTE**:
Due to the way the missing cpu time instrumentation warning works, Thread.start/fork will still trigger it, even if Threads are
instrumented correctly.

Since https://github.com/DataDog/dd-trace-rb/pull/1310 is still under review, for now I'm leaving it as a `FIXME` (and marking this PR as a draft), but this should be fixed before merging.

(I can't think of a 100% accurate way of fixing it, so possibly the fix will be along the lines of "track how long the thread
has lived, and if it was just created, don't warn yet -- wait a few seconds before considering it as missing instrumentation").

Hopefully at some point https://github.com/DataDog/dd-trace-rb/pull/1333 will become a better solution to all this messing around.

### CODE REVIEW

**Comment by ivoanjo** [comment]
Note: CI is broken on Ruby 3.0 because of the issue fixed in https://github.com/DataDog/dd-trace-rb/pull/1345 . I'll rebase this PR after that one gets merged to add `ruby2_keywords` to `start`, which fixes the breaking test.

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (marcotc) [inline lib/ddtrace/profiling/collectors/stack.rb:192]: Ugh, over the lifetime of an application this will likely get triggered a few times.

Would it make sense to give a thread one opportunity to fulfill `@warned_about_missing_cpu_time_instrumentation`, something like setting a sentinel value in the thread object that tracks "we've given you one chance to run update_native_ids, but next time I check, I'll warn if you didn't". But this suggestion has the side-effect of missing the warning on short-lived threads.

What do you think?

**Review by marcotc — COMMENTED** [review]
Looks very good!
I left a comment for discussion, but I don't think it blocks this specific PR.

File: lib/ddtrace/profiling/collectors/stack.rb
Comment (ivoanjo) [reply lib/ddtrace/profiling/collectors/stack.rb:192]: Agreed. It's a shame that it ends up making the detection miss a few cases, but hopefully we're mostly done with the whack-a-mole game so it's not going to come up often.

I'll wait until #1310 gets merged to add this change, as there's already quite a bit of branch dependencies around, and after that I'll unmark as draft.

**Comment by ivoanjo** [comment]
Update on this one:
* I've rebased this branch on top of the current `feature/profiling` state
* I've integrated the Ruby 3 fixes
* I've hid the instrumentation missing message, rather than fixing it. See the commit message in fa8c193 for the rationale behind my decision.

Thus, it's ready for review again! @marcotc or @delner can I bother one of you for a second look? :)

**Review by marcotc — APPROVED** [review]
:shipit:

**Pull request merged**
